File: packages/shared/lib/cycle.ts
import type { CycleDay, CyclePhase } from '../types'

// ============================
// 배란일·가임기 계산 로직
// 웹·앱 공통 사용 — 절대 중복 구현하지 말 것
// ============================

/**
 * 마지막 생리 시작일과 평균 주기로 향후 60일간의 사이클 데이터 생성
 */
export function calculateCycleDays(
  lastPeriodStart: Date,
  cycleLength: number = 28,
  periodLength: number = 5,
  daysAhead: number = 60
): CycleDay[] {
  const days: CycleDay[] = []

  // 배란일 = 다음 생리 예정일 - 14일 (황체기 고정)
  const ovulationDay = cycleLength - 14

  for (let i = 0; i < daysAhead; i++) {
    const date = new Date(lastPeriodStart)
    date.setDate(date.getDate() + i)

    // 현재 사이클 내 몇 번째 날인지 (1-based)
    const dayOfCycle = ((i % cycleLength) + 1)

    const isMenstruation = dayOfCycle <= periodLength
    const isOvulation = dayOfCycle === ovulationDay
    // 배란일 전후 5일 = 가임기 (배란일 -3 ~ +1)
    const isFertileWindow = dayOfCycle >= ovulationDay - 3 && dayOfCycle <= ovulationDay + 1

    const phase: CyclePhase = isMenstruation
      ? 'menstrual'
      : dayOfCycle < ovulationDay - 3
        ? 'follicular'
        : isOvulation || isFertileWindow
          ? 'ovulation'
          : 'luteal'

    days.push({
      date: formatDate(date),
      phase,
      isOvulation,
      isFertileWindow,
      isMenstruation,
      dayOfCycle,
    })
  }

  return days
}

/**
 * 다음 배란 예정일 반환
 */
export function getNextOvulationDate(
  lastPeriodStart: Date,
  cycleLength: number = 28
): Date {
  const ovulationDay = cycleLength - 14
  const next = new Date(lastPeriodStart)
  next.setDate(next.getDate() + ovulationDay - 1)
  return next
}

/**
 * 다음 생리 예정일 반환
 */
export function getNextPeriodDate(
  lastPeriodStart: Date,
  cycleLength: number = 28
): Date {
  const next = new Date(lastPeriodStart)
  next.setDate(next.getDate() + cycleLength)
  return next
}

/**
 * 오늘 기준 사이클 몇 번째 날인지
 */
export function getCurrentCycleDay(
  lastPeriodStart: Date,
  cycleLength: number = 28
): number {
  const today = new Date()
  const diffMs = today.getTime() - lastPeriodStart.getTime()
  const diffDays = Math.floor(diffMs / (1000 * 60 * 60 * 24))
  return (diffDays % cycleLength) + 1
}

function formatDate(date: Date): string {
  return date.toISOString().split('T')[0]
}

File: apps/mobile_flutter/lib/core/domain/cycle.dart
import '../models/cycle.dart';

/// Port of packages/shared/lib/cycle.ts — ovulation/fertile-window date math.
/// Keep this in exact sync with the TS original; do not reimplement independently.
class CycleCalc {
  CycleCalc._();

  static List<CycleDay> calculateCycleDays(
    DateTime lastPeriodStart, {
    int cycleLength = 28,
    int periodLength = 5,
    int daysAhead = 60,
  }) {
    final days = <CycleDay>[];
    final ovulationDay = cycleLength - 14;
    final startLocal = DateTime(
      lastPeriodStart.year,
      lastPeriodStart.month,
      lastPeriodStart.day,
    );

    for (var i = 0; i < daysAhead; i++) {
      final date = startLocal.add(Duration(days: i));
      final dayOfCycle = (i % cycleLength) + 1;

      final isMenstruation = dayOfCycle <= periodLength;
      final isOvulation = dayOfCycle == ovulationDay;
      final isFertileWindow =
          dayOfCycle >= ovulationDay - 3 && dayOfCycle <= ovulationDay + 1;

      final String phase = isMenstruation
          ? 'menstrual'
          : dayOfCycle < ovulationDay - 3
          ? 'follicular'
          : (isOvulation || isFertileWindow)
          ? 'ovulation'
          : 'luteal';

      days.add(
        CycleDay(
          date: _formatDate(date),
          phase: phase,
          isOvulation: isOvulation,
          isFertileWindow: isFertileWindow,
          isMenstruation: isMenstruation,
          dayOfCycle: dayOfCycle,
        ),
      );
    }

    return days;
  }

  static DateTime getNextOvulationDate(
    DateTime lastPeriodStart, {
    int cycleLength = 28,
  }) {
    final ovulationDay = cycleLength - 14;
    final start = DateTime(
      lastPeriodStart.year,
      lastPeriodStart.month,
      lastPeriodStart.day,
    );
    return start.add(Duration(days: ovulationDay - 1));
  }

  static DateTime getNextPeriodDate(
    DateTime lastPeriodStart, {
    int cycleLength = 28,
  }) {
    final start = DateTime(
      lastPeriodStart.year,
      lastPeriodStart.month,
      lastPeriodStart.day,
    );
    return start.add(Duration(days: cycleLength));
  }

  static int getCurrentCycleDay(
    DateTime lastPeriodStart, {
    int cycleLength = 28,
  }) {
    final now = DateTime.now();
    final diffDays = now.difference(lastPeriodStart).inDays;
    return (diffDays % cycleLength) + 1;
  }

  static String _formatDate(DateTime date) {
    final y = date.year.toString().padLeft(4, '0');
    final m = date.month.toString().padLeft(2, '0');
    final d = date.day.toString().padLeft(2, '0');
    return '$y-$m-$d';
  }
}
